Tests/Services/HealthTests/HealthTests.swift

```diff
@@ -0,0 +1,236 @@
+              status: statuses[i + 1]
+            )
+
+            try healthProvider.updateService(
+              descriptor: ignoredServiceDescriptor,
+              status: .notServing
+            )
+          }
+        }
+      }
+    }
+  }
+
+  func testWatchOnUnknownService() async throws {
+    try await withHealthClient { (healthClient, healthProvider) in
+      try healthProvider.updateService(
+        descriptor: ServiceDescriptor(package: "package.to.be.ignored", service: "IgnoredService"),
+        status: .serving
+      )
+
+      let testServiceDescriptor = ServiceDescriptor(package: "test.package", service: "TestService")
+
+      var message = Grpc_Health_V1_HealthCheckRequest()
+      message.service = testServiceDescriptor.fullyQualifiedService
+
+      try await healthClient.watch(request: ClientRequest.Single(message: message)) { response in
+        var responseStreamIterator = response.messages.makeAsyncIterator()
+        var next = try await responseStreamIterator.next()!
+
+        XCTAssertEqual(next.status, .serviceUnknown)
+
+        try healthProvider.updateService(
+          descriptor: testServiceDescriptor,
+          status: .notServing
+        )
+
+        next = try await responseStreamIterator.next()!
+
+        XCTAssertEqual(next.status, .notServing)
+      }
+    }
+  }
+
+  func testMultipleWatchOnTheSameService() async throws {
+    try await withHealthClient { (healthClient, healthProvider) in
+      let testServiceDescriptor = ServiceDescriptor(package: "test.package", service: "TestService")
+
+      let receivedStatuses1 = AsyncStream<Grpc_Health_V1_HealthCheckResponse.ServingStatus>
+        .makeStream()
+      let receivedStatuses2 = AsyncStream<Grpc_Health_V1_HealthCheckResponse.ServingStatus>
+        .makeStream()
+
+      let statusesToBeSent: [ServingStatus] = [
+        .serving,
+        .notServing,
+        .serving,
+        .serving,
+        .notServing,
+      ]
+
+      @Sendable func runWatch(
+        continuation: AsyncStream<Grpc_Health_V1_HealthCheckResponse.ServingStatus>.Continuation
+      ) async throws {
+        var message = Grpc_Health_V1_HealthCheckRequest()
+        message.service = testServiceDescriptor.fullyQualifiedService
+
+        try await healthClient.watch(request: ClientRequest.Single(message: message)) { response in
+          var responseStreamIterator = response.messages.makeAsyncIterator()
+
+          // Since responseStreamIterator.next() will never be nil (as the "watch" response stream
+          // is always open), the iteration cannot be based on when responseStreamIterator.next()
+          // is nil. Else, the iteration infinitely awaits and the test never finishes. Hence, it is
+          // based on the expected number of statuses to be received.
+          for _ in 0 ..< statusesToBeSent.count {
+            try await continuation.yield(responseStreamIterator.next()!.status)
+          }
+        }
```

Comment:
We don't need a continuation here, we can just get the first N elements from the response stream and collect them into an array and then return that from the RPC:

```swift
try await response.messages.prefix(N).reduce(into: []) { $0.append($1) }
```

Breaking that down:

```swift
// Get an async sequence which only returns the first N messages from `response.messages`
let prefix = response.messages.prefix(N) 

// Wait for the prefix sequence to end (i.e. collect all N elements) and store them in an array:
let reduced = prefix.reduce(into: []) { elements, nextElement in 
  elements.append(nextElement)
}
```

Reply:
We can then reduce this test to:

```swift
try await withThrowingTaskGroup(of: [Status].self) { group in
  let concurrentStreams = 2
  for _ in 0 ..< concurrentStreams {
    group.addTask {
      try await watch()
    }
  }

  for status in statuses {
    // update status
  }

  for try await next in group {
    XCTAssertEqual(next, expected)
  }
}
```